## Option Pricing using Monte Carlo Simulation

Option pricing using Monte Carlo simulation and geometric brownian motion. Compared with Black Scholes method.

Will calculate CALL and PUT Payoffs.

In [1]:
# Libraries

import numpy as np
from scipy.stats import norm

In [3]:
# Variables:

S = 385                 # Spot price
K = 350                 # Strike price
T = 0.083               # Time to expiration (Years)
r = 0.04                # Discount rate (%)     
sigma = 0.15            # Volatility of stock(%)


# Variables for Monte Carlo:

steps = 30              # Number of days option runs for
simulations = 5         # Number of simulations 

In [18]:
# Brownian motion for daily prices:
dt = T / steps

In [19]:
# Drift:
drift = (r-(sigma**2)/2) * dt

In [20]:
# Time:
a = sigma * np.sqrt(dt)

In [21]:
# Random variable: Creating random variables for each day of stock.

# Using standard normal distribution with mean=0, variance=1, (Nsimulation for Nsteps) 
x = np.random.normal(0,1,(simulations, steps))      
print(x)

[[-6.74312192e-02 -3.43892628e-01 -1.69061136e+00  9.67148798e-01
  -4.00918258e-02 -1.04829524e+00  7.96421786e-02 -1.28925555e+00
  -1.54456507e+00 -2.70125505e-01 -5.61108555e-01 -5.03677348e-01
   3.56090012e-01  5.10415375e-01  5.90981535e-01  1.02727485e+00
  -8.22120128e-01  1.75259177e-01 -5.10821637e-01  1.27803654e-01
   6.91000382e-01 -1.05484710e-01 -1.55153210e-02  7.91593728e-01
  -1.91421683e+00  8.64733899e-01 -1.08947033e+00  3.96925217e-01
  -1.26502686e+00  6.08674466e-01]
 [-1.06098354e+00  7.53675316e-02 -6.89153863e-01  5.90547647e-01
  -1.60520879e+00 -1.90901975e-01 -6.69867499e-01 -9.30113392e-01
   6.28891650e-01 -4.01132701e-01  8.34770795e-01 -5.32992761e-01
   1.94233873e+00  6.06715702e-01  1.08395398e-01  1.03003241e+00
   5.88096620e-01  1.08436348e+00 -3.46614255e-01 -1.23747384e+00
   1.89056282e-01  3.41193643e-01  6.91654807e-01 -5.49314328e-01
  -1.38932411e+00 -3.63767575e-01 -3.24310597e-02 -1.18050064e+00
   3.84910040e-01 -1.23637612e+00]
 [-2.4

Above we can see that for each of the 5 arrays (simulations), there are 30 elements (steps). Each element represents one day.

In [22]:
# Matrix of stock price values

Smat = np.zeros((simulations, steps))       # Matrix initially loaded with zero values
Smat[:,0]+= S                               # put 385 (Spot price) on the first element of each of the 5 arrays
Smat

array([[385.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [385.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [385.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [385.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [385.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0., 

Above we can see that the each of the 5 arrays contain a value of 385 (Spot price) at the 1st element, with 0's afterwards. These 0's will be exchanged for a value based on the random variables generated prevoiusly.

In [23]:
# Brownian Motion: (from the 2nd element onwards, we have the Brownian motion values)
# Each simulation represents a different Brownian Motion path.

for i in range(1, Nsteps):
    Smat[:,i] += Smat[:,i-1] * np.exp(drift + a*x[:,i])
Smat

array([[385.        , 383.98734983, 378.92960608, 381.86252776,
        381.77212201, 378.65765924, 378.92580888, 375.1207289 ,
        370.60657186, 369.84697318, 368.24254249, 366.81124552,
        367.8725118 , 369.3863421 , 371.14224588, 374.19236299,
        371.8026222 , 372.34671182, 370.87855654, 371.2822532 ,
        373.34167024, 373.06075474, 373.04476116, 375.41178859,
        369.81399093, 372.37534185, 369.21757145, 370.40511928,
        366.75570472, 368.55054776],
       [385.        , 385.25964738, 383.20102661, 385.02128141,
        380.2060213 , 379.66398731, 377.69273469, 374.96101653,
        376.8561229 , 375.69518476, 378.20784838, 376.65069054,
        382.49766299, 384.36360447, 384.72306284, 387.89322887,
        389.72823549, 393.10812034, 392.06572468, 388.28729958,
        388.89784467, 389.97717553, 392.14231375, 390.47749794,
        386.25134222, 385.1749964 , 385.10708253, 381.56719954,
        382.7581821 , 379.07273898],
       [385.        , 390.8268

Above we can see that we have 5 different Brownian Motion paths (simulations) for the stock price.

### Calulate Payoffs for CALL and PUT:

In [24]:
# CALL payoff values:

c = (Smat[:,-1]) - K        # Call = 'last value of Smat' subtract 'Strike Price'

for i in range(len(c)):
    if c[i] < 0:
        c[i] = 0            # where values are less than 1, make these values equal zero
    else:
        c[i] = c[i]         # otherwise take value as it is                
c

array([18.55054776, 29.07273898, 43.93941014, 14.65100066, 52.91966332])

In [25]:
# PUT payoff values:

p = K - (Smat[:,-1])        # Put = 'Strike Price' subtract the 'Spot price' 
for i in range(len(p)):
    if p[i] < 0:
        p[i] = 0            # where values are less than 1, make all values equal zero
    else:
        p[i] = p[i]         # otherwise take value as it is                
p

array([0., 0., 0., 0., 0.])

In [26]:
# Take the average of payoff simulations:

call_payoff = np.mean(c)
print('Call payoff =', call_payoff)

put_payoff = np.mean(p)
print('Put payoff =', put_payoff)

Call payoff = 31.82667216994372
Put payoff = 0.0


### Discount price back to present day:

In [27]:
call = call_payoff*np.exp(-r*T)
put = put_payoff*np.exp(-r*T)

In [28]:
print('Call payoff (Discounted to present day) =', call)
print('Put payoff (Discounted to present day) =', put)

Call payoff (Discounted to present day) = 31.721182827543352
Put payoff (Discounted to present day) = 0.0


We can see that the call payoff has value of 31.72, whereas the put payoff has a value of 0.0. This is because for these given spot and strike prices they only work in favor of a call option. Hence a Put option at these prices would not be executed. 

## Black Scholes Method:

In [7]:
# Black Scholes d1 & d2:

d1 = (np.log(S/K) + (r + sigma**2/2)*T) / sigma*np.sqrt(T)
d2 = (np.log(S/K) + (r - sigma**2/2)*T) / sigma*np.sqrt(T)

In [12]:
# Black Scholes Call option price:

call_bs = (norm.cdf(d1,0,1) * S) - (norm.cdf(d2,0,1) * K*np.exp(-r*T))
call_bs

21.31220748950824

In [11]:
call_discounted = call_bs*np.exp(-r*T)
print('Call payoff (Discounted to present day) =', call_discounted)

Call payoff (Discounted to present day) = 21.24156828660434


Using the same variables, the values have have been entered into the Black Scholes model. The call payoff has a value of 21.24.